File: scripts/scenarios/lib/validators/component-detector.js
'use strict';

/**
 * Component detector - Identifies which components (API, UI) exist in a scenario
 * @module scenarios/lib/validators/component-detector
 */

const fs = require('node:fs');
const path = require('node:path');

/**
 * Detect which components exist in the scenario
 * @param {string} scenarioRoot - Scenario root directory
 * @returns {Set<string>} Set of component types (API, UI)
 */
function detectScenarioComponents(scenarioRoot) {
  const components = new Set();

  // Check for API component
  const apiDir = path.join(scenarioRoot, 'api');
  if (fs.existsSync(apiDir)) {
    const hasGoFiles = fs.readdirSync(apiDir).some(f => f.endsWith('.go'));
    if (hasGoFiles) components.add('API');
  }

  // Check for UI component
  const uiDir = path.join(scenarioRoot, 'ui');
  if (fs.existsSync(uiDir)) {
    const pkgJson = path.join(uiDir, 'package.json');
    if (fs.existsSync(pkgJson)) components.add('UI');
  }

  return components;
}

/**
 * Get applicable validation layers for scenario based on components
 * @param {Set<string>} components - Detected components
 * @returns {Set<string>} Applicable layer names (API, UI, E2E)
 */
function getApplicableLayers(components) {
  const layers = new Set(['E2E']);  // E2E always applicable

  if (components.has('API')) {
    layers.add('API');
  }
  if (components.has('UI')) {
    layers.add('UI');
  }

  return layers;
}

module.exports = {
  detectScenarioComponents,
  getApplicableLayers
};

File: scenarios/scenario-completeness-scoring/api/pkg/validators/component_detector.go
package validators

import (
	"os"
	"path/filepath"
	"strings"
)

// ScenarioComponents represents which components exist in a scenario
type ScenarioComponents struct {
	HasAPI bool
	HasUI  bool
}

// DetectScenarioComponents detects which components (API, UI) exist in a scenario
func DetectScenarioComponents(scenarioRoot string) ScenarioComponents {
	components := ScenarioComponents{}

	// Check for API component
	apiDir := filepath.Join(scenarioRoot, "api")
	if info, err := os.Stat(apiDir); err == nil && info.IsDir() {
		entries, err := os.ReadDir(apiDir)
		if err == nil {
			for _, entry := range entries {
				if strings.HasSuffix(entry.Name(), ".go") {
					components.HasAPI = true
					break
				}
			}
		}
	}

	// Check for UI component
	uiDir := filepath.Join(scenarioRoot, "ui")
	if info, err := os.Stat(uiDir); err == nil && info.IsDir() {
		pkgJSON := filepath.Join(uiDir, "package.json")
		if _, err := os.Stat(pkgJSON); err == nil {
			components.HasUI = true
		}
	}

	return components
}

// GetApplicableLayers returns which validation layers apply to a scenario
func GetApplicableLayers(components ScenarioComponents) map[string]bool {
	layers := map[string]bool{
		"E2E": true, // E2E always applicable
	}

	if components.HasAPI {
		layers["API"] = true
	}
	if components.HasUI {
		layers["UI"] = true
	}

	return layers
}

// ComponentsToSet converts ScenarioComponents to a set of component names
func (c ScenarioComponents) ToSet() map[string]bool {
	set := make(map[string]bool)
	if c.HasAPI {
		set["API"] = true
	}
	if c.HasUI {
		set["UI"] = true
	}
	return set
}
